zquery/shared/src/main/scala/zio/zquery/DataSource.scala

```diff
@@ -111,14 +118,38 @@ trait DataSource[-R, -A] { self =>
 
 object DataSource {
 
+  /**
+   * A data source that batches queries that can be performed in parallel in
+   * batches but does not pipeline queries that must be performed sequentially.
+   */
+  trait Batched[-R, -A] extends DataSource[R, A] {
+    def run(requests: Iterable[A]): ZIO[R, Nothing, CompletedRequestMap]
+    final def runAll(requests: Iterable[Iterable[A]]): ZIO[R, Nothing, CompletedRequestMap] =
+      ZIO.foreach(requests)(run).map(_.foldLeft(CompletedRequestMap.empty)(_ ++ _))
+  }
+
+  object Batched {
+
+    /**
+     * Constructs a data source from a function taking a collection of requests
+     * and returning a `CompletedRequestMap`.
+     */
+    def apply[R, A](name: String)(f: Iterable[A] => ZIO[R, Nothing, CompletedRequestMap]): DataSource[R, A] =
```

Comment:
I think it's a little odd to use multiparameter lists with constructors called `apply` (nothing wrong with it, just violates the normal `apply` convention).

Reply:
Maybe we call it `make`?

Reply:
Sounds great!